### PR TITLE
feat(taskframework): add process run model for task running

### PR DESCRIPTION
**Pull request by krihy**
#### What type of PR is this?

type-feature
module-taskframework
#### What this PR does / why we need it:
Task run in process model,  when user set  `odc.task-framework.run-mode` = 'PROCESS', new task will be started in a new process,  this is default model for task-framework. process will be destroyed when task status is terminate.

#### Which issue(s) this PR fixes:

1. add `  <groupId>net.java.dev.jna</groupId> <artifactId>jna</artifactId>` for Windows operate process
2. interface`Kernel32` will load native dynamic library to get process id, add `killProcessByPid` and `isProcessRunning` in SystemUtils  for kill and find process by id.
3. modify task log4j config,  task executor log will put set in path `${sys:odc.log.directory:-./log/task}`, if in k8s model, default mount path is `/data/logs/odc-task/runtime/task` 
4. add `DestroyExecutorJob` to destroy executor when job status is terminate

#### Special notes for your reviewer:

note: current pr is tested on *inx platform,  this should be test again on Windows.

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-common/src/main/java/com/oceanbase/odc/common/util/SystemUtils.java [hunk 1]
```diff
@@ -146,13 +151,103 @@ public static long getProcessPid(Process process) {
...
+
+    public static boolean isProcessRunning(long pid) {
+        if (-1 == pid) {
+            throw new IllegalArgumentException("query process by illegal argument pid: " + pid);
+        }
+        String command = "";
+        if (isOnWindows()) {
+            // tasklist exit code is always 0. Parse output
+            // findstr exit code 0 if found pid, 1 if it doesn't
+            command = "cmd.exe /c \"tasklist /FI \"PID eq " + pid + "\" | findstr " + pid + "\"";
```
Comment (smallsheeeep) [inline server/odc-common/src/main/java/com/oceanbase/odc/common/util/SystemUtils.java:202]: There seems to be a problem with the splicing of this command, please confirm it.

Comment (krihy) [reply server/odc-common/src/main/java/com/oceanbase/odc/common/util/SystemUtils.java:202]: looks normally,  find process which pid equal 1 should like: cmd.exe /c \"tasklist /FI \"PID eq  1 \" | findstr 1 + \"

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/ProcessJobCaller.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    protected void doStop(JobIdentity ji) throws JobException {}
+
+    @Override
+    protected void doDestroy(ExecutorIdentifier identifier) throws JobException {
+        if (SystemUtils.getLocalIpAddress().equals(identifier.getHost())) {
+            if (SystemUtils.isProcessRunning(Long.parseLong(identifier.getExecutorName()),
+                    Long.parseLong(identifier.getNamespace()))) {
+                log.info("Found process {}, kill it.", identifier.getExecutorName());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/ProcessJobCaller.java:72]: use comma separated style for log output.

`Found process, will kill it , process={}`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/constants/JobEnvKeyConstants.java [hunk 3]
```diff
@@ -43,17 +45,19 @@ public class JobEnvKeyConstants {
 
     public static final String ODC_OBJECT_STORAGE_CONFIGURATION = "ODC_OBJECT_STORAGE_CONFIGURATION";
 
-    public static final String DATABASE_HOST = "DATABASE_HOST";
+    public static final String ODC_DATABASE_HOST = "ODC_DATABASE_HOST";
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/constants/JobEnvKeyConstants.java:48]: JOBMETA_DATABASE_HOST

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java [hunk 4]
```diff
@@ -81,28 +86,41 @@ private void init(String[] args) {
...
 
-        log.info("JobContextProvider init success: {}", jobContextProvider.getClass().getSimpleName());
+        String runMode = SystemUtils.getEnvOrProperty(JobEnvKeyConstants.ODC_TASK_RUN_MODE);
+        log.info("ODC TASK RUN MODE is {}.", runMode);
+
         taskExecutor = ThreadPoolTaskExecutor.getInstance();
         log.info("Task executor init success: {}", taskExecutor.getClass().getSimpleName());
         log.info("Task executor ip is {}.", SystemUtils.getLocalIpAddress());
-        log.info("Task executor port is {}.", JobUtils.getPort());
+        log.info("Task executor port is {}.", JobUtils.getExecutorPort());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java:105]: log.info(Task executor init success, executorName={}, ip={}, port={});

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java [hunk 5]
```diff
@@ -81,28 +86,41 @@ private void init(String[] args) {
...
+    }
+
+    private void decryptIntercept() {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java:108]: decryptEnvironments

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java [hunk 6]
```diff
@@ -81,28 +86,41 @@ private void init(String[] args) {
...
+        System.getProperties().forEach((key, value) -> {
+            allProperties.put((String) key, (String) value);
+        });
+        new JobDecryptInterceptor().intercept(allProperties);
     }
 
     private void trace(long taskId) {
-        String taskLogFile = "log4j2-task-executor.xml";
+        TraceContextHolder.trace();
+        // mock userId
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java:118]: userId should be set correctly

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/EmbedServer.java [hunk 7]
```diff
@@ -112,9 +114,18 @@ public void initChannel(SocketChannel channel) throws Exception {
...
+                        future = bootstrap.bind(JobUtils.getExecutorPort().get()).sync();
+                        port = JobUtils.getExecutorPort().get();
+                    } else {
+                        future = bootstrap.bind(0).sync();
+                        InetSocketAddress localAddress = (InetSocketAddress) future.channel().localAddress();
+                        // save port to system properties
+                        JobUtils.setExecutorPort(localAddress.getPort());
+                        port = localAddress.getPort();
+                    }
                     log.info(">>>>>>>>>>> odc-job remoting server start success, nettype = {}, port = {}",
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/EmbedServer.java:129]: remove '>>>>>>>>>>> '

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/TaskMonitor.java [hunk 8]
```diff
@@ -68,27 +68,31 @@ public void monitor() {
...
             try {
-                reportTaskResult();
+                if (JobUtils.getExecutorPort().isPresent()) {
+                    reportTaskResult();
+                }
             } catch (Throwable e) {
                 log.warn("Update task info failed, id: {}", getJobId(), e);
             }
-        }, 1, JobConstants.REPORT_TASK_INFO_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        }, 5, JobConstants.REPORT_TASK_INFO_INTERVAL_SECONDS, TimeUnit.SECONDS);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/TaskMonitor.java:81]: extract magic number into constant fields

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/TaskReporter.java [hunk 9]
```diff
@@ -43,6 +46,10 @@ public TaskReporter(List<String> hostUrls) {
 
 
     public <T> boolean report(String url, T result) {
+        String reportEnabledValue = SystemUtils.getEnvOrProperty(JobEnvKeyConstants.REPORT_ENABLED);
+        if (reportEnabledValue != null && Objects.equals(Boolean.valueOf(reportEnabledValue), Boolean.FALSE)) {
+            return true;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/server/TaskReporter.java:51]: judge report enabled in method caller

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/task/BaseTask.java [hunk 10]
```diff
@@ -55,7 +55,7 @@ public void start(JobContext context) {
             doStart(context);
             updateStatus(JobStatus.DONE);
         } catch (Throwable e) {
-            log.info("Task failed, id: {}, details: {}", context.getJobIdentity().getId(), e);
+            log.info("Task failed, id:" + context.getJobIdentity().getId() + ", details: ", e);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/task/BaseTask.java:58]: log style

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+@Slf4j
+@DisallowConcurrentExecution
+public class DestroyExecutorJob implements Job {
+
+    private JobConfiguration configuration;
+
+    @Override
+    public void execute(JobExecutionContext context) throws JobExecutionException {
+        configuration = JobConfigurationHolder.getJobConfiguration();
+        if (configuration == null) {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java:49]: throw exceptoin if configuration==null

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                destroyExecutor(taskFrameworkService, a);
+            } catch (Throwable e) {
+                log.warn("Try to start job {} failed: ", a.getId(), e);
+            }
+        });
+    }
+
+    private void destroyExecutor(TaskFrameworkService taskFrameworkService, JobEntity oldEntity) {
+        getConfiguration().getTransactionManager().doInTransactionWithoutResult(() -> {
+            JobEntity newEntity = taskFrameworkService.findWithPessimisticLock(oldEntity.getId());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java:69]: refine naming for newEntity

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        // scan terminate job
+        TaskFrameworkService taskFrameworkService = configuration.getTaskFrameworkService();
+        TaskFrameworkProperties taskFrameworkProperties = configuration.getTaskFrameworkProperties();
+        Page<JobEntity> jobs = taskFrameworkService.findTerminalJob(0,
+                taskFrameworkProperties.getSingleFetchDestroyExecutorJobRows());
+        jobs.forEach(a -> {
+            try {
+                destroyExecutor(taskFrameworkService, a);
+            } catch (Throwable e) {
+                log.warn("Try to start job {} failed: ", a.getId(), e);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java:62]: wrong log message

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/StdJobScheduler.java [hunk 14]
```diff
@@ -154,6 +156,13 @@ private void initDoCancelingJob() {
                 DoCancelingJob.class);
     }
 
+    private void initDestroyExecutorJob() {
+        String key = "destroyExecutorJob";
+        initCronJob(key, key + "Group",
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/StdJobScheduler.java:161]: may use 'odcJobMonitoring' as trigger group, extract to constants class

**Review by yizhouxw — COMMENTED** [review]
another notice  point: check QuartzConfiguration, if odcScheduler and taskframeworkSchduler share the same thread pool

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/ExecutorIdentifierParser.java [hunk 15]
```diff
@@ -27,12 +27,12 @@
  */
 public class ExecutorIdentifierParser {
 
-    public static ExecutorIdentifier parser(String identifierString) throws JobException {
+    public static ExecutorIdentifier parser(String identifierString) {
         UriComponents uriComponents = UrlUtils.getUriComponents(identifierString);
```
[the remark this answers is not in the document]
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/ExecutorIdentifierParser.java:31]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java (hunk 6 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/TaskApplication.java:113]: okey, done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/task/BaseTask.java (hunk 10 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/executor/task/BaseTask.java]: okey, done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java (hunk 11 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java (hunk 13 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/schedule/daemon/DestroyExecutorJob.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/DefaultExecutorIdentifier.java [hunk 16]
```diff
@@ -32,7 +32,7 @@ public class DefaultExecutorIdentifier implements ExecutorIdentifier {
 
     public static final String DEFAULT_PROTOCOL = "http";
     public static final String DEFAULT_HOST = "odc";
```
[the remark this answers is not in the document]
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/caller/DefaultExecutorIdentifier.java:33]: UriComponentsBuilder fromUriString(url) will parser url to ExecutorIdentifier, if hostname is null, will cause exception, ExecutorIdentifier should be designed again

File: script/start-job.sh [hunk 17]
```diff
@@ -0,0 +1,193 @@
+#! /bin/bash
+# Start up script for odc-server
```
[the remark this answers is not in the document]
Comment (krihy) [reply script/start-job.sh]: done

File: script/start-job.sh [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+# constants
+script_source=$(readlink -f $0)
+bin_directory=$(dirname $script_source)
+install_directory=$(dirname $bin_directory)
+app_log_config_file="${install_directory}/conf/log4j2-task.xml"
+current_work_directory="$(pwd)"
+default_server_port=8989
+gc_basic_options="-XX:+UseG1GC -XX:+PrintAdaptiveSizePolicy -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps"
+gc_log_options="-Xloggc:${install_directory}/log/task/gc.log -XX:+UseGCLogFileRotation -XX:GCLogFileSize=50M -XX:NumberOfGCLogFiles=5"
+default_heap_options="-XX:MaxRAMPercentage=100.0 -XX:InitialRAMPercentage=100.0"
```
[the remark this answers is not in the document]
Comment (krihy) [reply script/start-job.sh]: ok , MaxRAMPercentage change to 60 as odc setting

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/SpringTaskFrameworkProperties.java [hunk 19]
```diff
@@ -35,9 +35,10 @@
 @ConfigurationProperties(prefix = "odc.task-framework")
 public class SpringTaskFrameworkProperties implements TaskFrameworkProperties {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/SpringTaskFrameworkProperties.java:36]: DefaultTaskFrameworkProperties

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/JobUtils.java [hunk 20]
```diff
@@ -54,39 +61,75 @@ public static String toJson(Object obj) {
...
+        if ((osc = System.getProperty(JobEnvKeyConstants.ODC_OBJECT_STORAGE_CONFIGURATION)) != null) {
+            ObjectStorageConfiguration storageConfig = JsonUtils.fromJson(osc, ObjectStorageConfiguration.class);
             return Optional.of(storageConfig);
         }
         return Optional.empty();
     }
+
+    public static Long getUserId() {
+        String userId = SystemUtils.getEnvOrProperty(JobEnvKeyConstants.ODC_EXECUTOR_USER_ID);
+        return StringUtils.isNotBlank(userId) ? Long.parseLong(userId) : 1L;
```
[the remark this answers is not in the document]
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/util/JobUtils.java]: add valid not blank for JobEnvKeyConstants.ODC_EXECUTOR_USER_ID

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/SpringTaskFrameworkProperties.java (hunk 19 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/SpringTaskFrameworkProperties.java]: done

File: script/start-job.sh [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+        spacev_java_agent_options="${SPACEV_JAVA_AGENT}"
+        echo "SPACEV_JAVA_AGENT is set"
+    fi
+    local log_options="-Dlog4j.configurationFile=${app_log_config_file} -Dodc.log.directory=${app_log_directory}"
+    local work_dir_options="-Duser.dir=${ODC_WORK_DIR:-${current_work_directory}}"
+    local plugin_options="-Dplugin.dir=${plugin_directory}"
+
+    app_options="${log_options} ${work_dir_options} ${plugin_options}"
+
+    local listen_port_args="--server.port=${server_port}"
```
[the remark this answers is not in the document]
Comment (krihy) [reply script/start-job.sh]: done

**Review by yizhouxw — APPROVED** [review]
lgtm

**Pull request merged**